LFS-1074: CARDS should not start if it cannot bind to a port

IntegralProgrammer opened the pull request:
This pull request introduces the `start_cards.sh` script which should be used as a replacement for calling `java -jar` directly. Upon starting the CARDS Java process, this script checks if CARDS was able to bind to the specified TCP port (default `8080`) within a timeout. If it was able to bind, the script prints a success message. Otherwise, the script terminates, stops CARDS, and prints an error message.

To test:

- Try the various starting commands for CARDS under the **Run** section of _README.md_. The startup script should display a success message once CARDS binds to the specified TCP port. CARDS should operate normally hereafter.
- Try the various starting commands in an environment where CARDS will not be able to bind to a TCP port. For example, start listening with `netcat` on port `8080` and then attempt to start CARDS on port `8080`. The startup script should terminate and display a failure message.

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

Fix README.md

sashaandjic commented:
Tried to run another instance on 8080 and got the failure message as expected.

I am not sure about this:
 
File "Utilities/HostConfig/check_tcp_listen.py", line 27, in <module>
    import psutil
ModuleNotFoundError: No module named 'psutil'
Traceback (most recent call last):
  File "Utilities/HostConfig/check_tcp_listen.py", line 27, in <module>
    import psutil

![image](https://user-images.githubusercontent.com/23243244/116764714-f05f7e00-a9ef-11eb-8c81-21c0317b0194.png)

sashaandjic commented:
I am unable to start an instance at any port.

marta- commented:
Just wondering: is python really needed? I would assume this functionality can be implemented in bash directly.

IntegralProgrammer commented:
@marta- The reason why I created the `check_tcp_listen.py` Python utility program is because all of the popular Linux command line utilities for doing this (eg. `netstat`, `fuser`, `lsof`, `ss`) all output formatted tabular data to the console and therefore I would have to `grep` to determine if a PID has bound to a port. If the output of one of these commands changed, then the startup script might break.

@sashaandjic On Ubuntu, you should be able to fix the `import psutil` error by installing the `python3-psutil` package (`sudo apt install python3-psutil`).

This raises another question - should we have some centralized list where we maintain a list of dependencies?

sashaandjic reviewed (approved):
tested

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

Modify check_tcp_listen.py so that it does not depend on the psutil Python module

IntegralProgrammer commented:
Updated `check_tcp_listen.py` so that it no longer depends on the `psutil` external Python module.

sashaandjic commented:
I was not able to start cards on already taken port.
However, when I ran in on 8080 port that was not taken, it started cards, but then after a minute it shut it down.
![image](https://user-images.githubusercontent.com/23243244/116947837-c904ed00-ac4b-11eb-8322-2e332e418908.png)

IntegralProgrammer commented:
> However, when I ran in on 8080 port that was not taken, it started cards, but then after a minute it shut it down.

Would you be able to try this again (start CARDS on port `8080` when port `8080` is available) but before CARDS terminates with the `Unable to bind to TCP port error`, send me the output of `cat /proc/net/tcp | grep ' 0A '` and ` cat /proc/net/tcp6 | grep ' 0A '`?

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

Move back to using the psutil Python module as it works on WSL while /proc/net/tcp does not

This reverts commit 658ab668cdbd0bf6a5980eafc57473710ef24650.

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

start_cards.sh will perform a more robust port-bind checking test if the psutil Python module is available, otherwise it will perform a less robust port-bind checking test where the port availability will be checked immediately before starting CARDS

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

Update README.md

IntegralProgrammer commented:
Updated `start_cards.sh` script so that it performs the ideal port-bind test if the `psutil` Python module is present, otherwise a simple port availability check is performed immediately before starting CARDS thus being susceptible to a very unlikely time-of-check-time-of-use bug.

@sashaandjic Please re-test with the `psutil` module installed as well as with the `psutil` module not installed.

sashaandjic commented:
Re-tested with the psutil module installed as well as with the psutil module removed.

marta- commented on start_cards.sh:
```diff
@@ -0,0 +1,83 @@
+if [ $PSUTIL_INSTALLED = false ]
+then
+  python3 Utilities/HostConfig/check_tcp_available.py --tcp_port $BIND_PORT || \
+    { echo -e "${TERMINAL_RED}Unable to bind to TCP port${TERMINAL_NOCOLOR}"; exit -1; }
```
Can we make it more visible? With all the stuff that gets printed out, the message might get lost even if it's red. Something like this:
```
**********************************************************************************
*                                                                                *
*    Message                                                                     *
*                                                                                *
**********************************************************************************

IntegralProgrammer replied:
Implemented in commit `5815921` :heavy_check_mark:

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

Print a border around the started / failed to start console messages

sashaandjic commented:
I never see the green message. If port in not taken, it starts the instance and after few minutes it dies. Each time it started I was able to created a couple of subjects.
Same results with psutil and without it.

IntegralProgrammer committed:
LFS-1074: CARDS should not start if it cannot bind to a port

Do not perform psutil based check in WSL as it always returns 'Unable to bind to TCP port'

IntegralProgrammer commented:
Updated this PR so that sub-optimal test (instead of `psutil` test) will always be performed on WSL, regardless of whether or not `psutil` is installed. @sashaandjic please re-test in WSL with `python3-psutil` installed and `python3-psutil` not installed. @marta- please re-test in Linux with `python3-psutil` installed.

sashaandjic reviewed (approved):
tested in WSL

- works well with and without python3-psutil; in both cases an instance starts with the following message
![image](https://user-images.githubusercontent.com/23243244/117494891-f353f480-af42-11eb-8abf-9f3ec7cae57e.png)

- if unable to bind to a port ( if port is already in use), the following message happens right away. Previously it was taking some time before getting this message.
![image](https://user-images.githubusercontent.com/23243244/117495058-27c7b080-af43-11eb-96c5-e5ed86f5cca8.png)

marta- reviewed (approved):
Works fine without and with psutil, all 3 messages are printed and visible.

Without psutil it immediately says it can't bind to port, while with psutil it takes a munite and a lot of messages printed out before it finally informs me the port is not available. I prefer the former.

sdumitriu commented on README.md:
```diff
@@ -29,15 +30,15 @@ To specify a different URL, use `-Dsling.url=https://lfs.server:8443/system/cons
 Installing larger vocabuleries may fail due to default limits imposed on XML documents by the JVM. In this case, the app should be started with an extra parameter:
 
-`java -Djdk.xml.entityExpansionLimit=0 -jar distribution/target/lfs-*.jar`
+`./start_cards.sh -Djdk.xml.entityExpansionLimit=0`
```
This should be in the script by default.
